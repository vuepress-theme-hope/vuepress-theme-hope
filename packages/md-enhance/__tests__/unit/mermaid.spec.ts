import MarkdownIt from "markdown-it";
import { describe, expect, it } from "vitest";
import { atou } from "vuepress-shared";

import { mermaid } from "../../src/node/markdown-it/index.js";

const demo = `flowchart TB
  c1-->a2
  subgraph one
  a1-->a2
  end
  subgraph two
  b1-->b2
  end
  subgraph three
  c1-->c2
  end
  one --> two
  three --> two
  two --> c2`;

describe("mermaid", () => {
  const markdownIt = MarkdownIt({ linkify: true }).use(mermaid);

  it("Should render ```mermaid", () => {
    const renderResult = markdownIt.render(`
\`\`\`mermaid
${demo}
\`\`\`
`);

    expect(renderResult).toMatch(
      /<Mermaid id="mermaid.*?" code=".*?"><\/Mermaid>/
    );
    expect(
      atou(
        /<Mermaid id="mermaid.*?" code="(.*?)"><\/Mermaid>/.exec(
          renderResult
        )?.[1] || ""
      )
    ).toMatch(demo);
    expect(renderResult).toMatchSnapshot();
  });

  it("Should not render", () => {
    expect(
      markdownIt.render(`
${demo}
`)
    ).toMatchSnapshot();

    expect(
      markdownIt.render(`
\`\`\`md
${demo}
\`\`\`
`)
    ).toMatchSnapshot();
  });

  it("Should render ```sequence", () => {
    const renderResult = markdownIt.render(`
\`\`\`sequence
Alice ->> Bob: Hello Bob, how are you?
Bob-->>John: How about you John?
Bob--x Alice: I am good thanks!
Bob-x John: I am good thanks!
Note right of John: Bob thinks a long<br/>long time, so long<br/>that the text does<br/>not fit on a row.

Bob-->Alice: Checking with John...
Alice->John: Yes... John, how are you?
\`\`\`
`);

    expect(renderResult).toMatchSnapshot();
  });

  it("Should render ```class", () => {
    const renderResult = markdownIt.render(`
\`\`\`class
class Square~Shape~{
  int id
  List~int~ position
  setPoints(List~int~ points)
  getPoints() List~int~
}

Square : -List~string~ messages
Square : +setMessages(List~string~ messages)
Square : +getMessages() List~string~
\`\`\`
`);

    expect(renderResult).toMatchSnapshot();
  });

  it("Should render ```state", () => {
    const renderResult = markdownIt.render(`
\`\`\`state
[*] --> Active

state Active {
  [*] --> NumLockOff
  NumLockOff --> NumLockOn : EvNumLockPressed
  NumLockOn --> NumLockOff : EvNumLockPressed
  --
  [*] --> CapsLockOff
  CapsLockOff --> CapsLockOn : EvCapsLockPressed
  CapsLockOn --> CapsLockOff : EvCapsLockPressed
  --
  [*] --> ScrollLockOff
  ScrollLockOff --> ScrollLockOn : EvScrollLockPressed
  ScrollLockOn --> ScrollLockOff : EvScrollLockPressed
}
\`\`\`
`);

    expect(renderResult).toMatchSnapshot();
  });

  it("Should render ```er", () => {
    const renderResult = markdownIt.render(`
\`\`\`er
CAR ||--o{ NAMED-DRIVER : allows
CAR {
  string registrationNumber
  string make
  string model
}
PERSON ||--o{ NAMED-DRIVER : is
PERSON {
  string firstName
  string lastName
  int age
}
\`\`\`
`);

    expect(renderResult).toMatchSnapshot();
  });

  it("Should render ```journey", () => {
    const renderResult = markdownIt.render(`
\`\`\`journey
title My working day
section Go to work
  Make tea: 5: Me
  Go upstairs: 3: Me
  Do work: 1: Me, Cat
section Go home
  Go downstairs: 5: Me
  Sit down: 5: Me
\`\`\`
`);

    expect(renderResult).toMatchSnapshot();
  });

  it("Should render ```gantt", () => {
    const renderResult = markdownIt.render(`
\`\`\`gantt
dateFormat  YYYY-MM-DD
title       Adding GANTT diagram functionality to mermaid
excludes    weekends
%% (\`excludes\` accepts specific dates in YYYY-MM-DD format, days of the week ("sunday") or "weekends", but not the word "weekdays".)

section A section
Completed task            :done,    des1, 2014-01-06,2014-01-08
Active task               :active,  des2, 2014-01-09, 3d
Future task               :         des3, after des2, 5d
Future task2              :         des4, after des3, 5d

section Critical tasks
Completed task in the critical line :crit, done, 2014-01-06,24h
Implement parser                    :crit, done, after des1, 2d
Create tests for parser             :crit, active, 3d
Future task in critical line        :crit, 5d
Create tests for renderer           :2d
Add to mermaid                      :1d

section Documentation
Describe gantt syntax               :active, a1, after des1, 3d
Add gantt diagram to demo page      :after a1  , 20h
Add another diagram to demo page    :doc1, after a1  , 48h

section Last section
Describe gantt syntax               :after doc1, 3d
Add gantt diagram to demo page      :20h
Add another diagram to demo page    :48h
\`\`\`
`);

    expect(renderResult).toMatchSnapshot();
  });

  it("Should render ```pie", () => {
    const renderResult = markdownIt.render(`
\`\`\`pie
title What Voldemort doesnâ€™t have?
  "FRIENDS" : 2
  "FAMILY" : 3
  "NOSE" : 45
\`\`\`
`);

    expect(renderResult).toMatchSnapshot();
  });

  it("Should render ```git-graph", () => {
    const renderResult = markdownIt.render(`
\`\`\`git-graph
commit id: "Normal"
commit
commit id: "Reverse" type: REVERSE
commit
commit id: "Highlight" type: HIGHLIGHT
commit
\`\`\`
`);

    expect(renderResult).toMatchSnapshot();
  });
});
