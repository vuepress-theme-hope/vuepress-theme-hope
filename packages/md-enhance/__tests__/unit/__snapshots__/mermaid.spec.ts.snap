// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mermaid > Should not render 1`] = `
"<p>flowchart TB
c1--&gt;a2
subgraph one
a1--&gt;a2
end
subgraph two
b1--&gt;b2
end
subgraph three
c1--&gt;c2
end
one --&gt; two
three --&gt; two
two --&gt; c2</p>
"
`;

exports[`mermaid > Should not render 2`] = `
"<pre><code class=\\"language-md\\">flowchart TB
  c1--&gt;a2
  subgraph one
  a1--&gt;a2
  end
  subgraph two
  b1--&gt;b2
  end
  subgraph three
  c1--&gt;c2
  end
  one --&gt; two
  three --&gt; two
  two --&gt; c2
</code></pre>
"
`;

exports[`mermaid > Should render \`\`\`class 1`] = `"<Mermaid id=\\"mermaid-0\\" code=\\"eJxdjTEOwjAMRfecwiMV4gLMjCAh9QQWRMESTUpsJoTPjutGbcUSxf8/+92eyHwiTBWHAOAj9K831qj9A8eoH4sBKAvQ3b9nYlGbFcbCJFSyxxzlWizm3RaYgs77tPTdesKab7BnNsIRDl6xVMpJYYjMmCJvib2JLi2fVf/05Ftx8y54M7eF8APCy1cA\\"></Mermaid>"`;

exports[`mermaid > Should render \`\`\`er 1`] = `"<Mermaid id=\\"mermaid-0\\" code=\\"eJx1jT0OwjAMhfeewhfIBdgqkoGBgILEboSJLPIjOUEMbe9OKnVoBgbLz89P7yPRjF4wDgDH0cE8K5UnsOPZaKXd6W4cHABDyN+yRaa2AUoVTh6EPDeJlXOyn/gg2X8jvqm785NCM5Y2V+NuF/sHyCtsS3S8F0upFmNXG3DncaqAftXL8ANM5kJr\\"></Mermaid>"`;

exports[`mermaid > Should render \`\`\`gantt 1`] = `"<Mermaid id=\\"mermaid-0\\" code=\\"eJyVU01v2zAMvfdXEAYKtIAzxGk2dLkFCToMWHfqZbeqEp0ItaVAorfm34+UP2A3adHqJJt6j4+P5E45ogsAowjvfKgVAfzhM7u/n223HCBLFUJ71sZYt4Mf698PD2Cs2gVVQ9k4TdY7VVk6AnmokWmsYSy+6KoxGAX7D/EZnYn8+/ISrh772CMorfFAEeIBtS2tTloiWDcSAmXSlnPsGMGXQHtMlHCVxcbx3+wafICsT5Pl8NQQOE/tUx9MGxSC7Mv1BeuImITDur/xv42vDxUSGiAVn7u601kZ7zCXG6sucljMi+VsXszm3/LhessMa6b6iydwYVApxBzMsBgxfM/hRvy6a6gJ56HDjaE3OaiSMHQ0X19BF+9AlyMo0zB05MMmWLJaVYlFGvXKDO6IeKn7Z5V1CCv55L4kc8aeLPfM8FMIanQEBxUi5z1zJgyDOnFYCtsE5HEAnggeEZ6Cc0QdQ2/viZcsfCp6iksGnuQJPEYYJplWSRFvwWjMz1XEL4uJs1uvG3FBdVO2xcipnxDS+kE8cujlrWlRbMXYllSeiGjB/R6yJIO1Z392XYGrFqUKAGmN9ENgipdiL2xvAHnQ9ZCyBS9v9+NyfqnIooed+Vg1bQWJ+uMVfEK1aPwPxhRkQQ==\\"></Mermaid>"`;

exports[`mermaid > Should render \`\`\`git-graph 1`] = `"<Mermaid id=\\"mermaid-0\\" code=\\"eJxLzyxxL0osyOBSUEjOz83NLFHITLFSUPLLL8pNzFGCi6JJB6WWpRYVpyoplFQWpFopBLmGuQYFu+JS7ZGZnpEDxCUw9R6e7h4+QByC0AEAoSMq3Q==\\"></Mermaid>"`;

exports[`mermaid > Should render \`\`\`journey 1`] = `"<Mermaid id=\\"mermaid-0\\" code=\\"eJxVzD0OgzAMhuG9p/gO0KWqumRtpU6ZegKrWGB+YpQYIW4PBAFifT77rXWIgacbYGItw08YNTYSShS0cuK/iQZ8FaZ5WxDw1DCMyeHl4DnbcjL0yUhicnge/NneHB4r3fEmu3Yr7Y5AoWPYE2f5J5aX3WYBdjXD\\"></Mermaid>"`;

exports[`mermaid > Should render \`\`\`mermaid 1`] = `"<Mermaid id=\\"mermaid-0\\" code=\\"eJxLy8kvT85ILCpRCHHiUlBINtTVtUs0ArKKS5PSixILMhTy81KB3ES4RGpeCrJ0SXk+kJsEkk7CJp1RlArSDzY4GaEAaKoCUASqHawKmV+eD+YBNQAAgWYsmw==\\"></Mermaid>"`;

exports[`mermaid > Should render \`\`\`pie 1`] = `"<Mermaid id=\\"mermaid-0\\" code=\\"eJwryEzlUlAoySzJSVUIz0gsUQjLz0lJzc0vKlFIyU8tznvUMLNEISOxLNUeqExBQcktyNPVzyVYScFKwQgq4ujr6RMJEjCGCPj5B7uCuCamXAB+DBmY\\"></Mermaid>"`;

exports[`mermaid > Should render \`\`\`sequence 1`] = `"<Mermaid id=\\"mermaid-0\\" code=\\"eJxtkL2OwjAQhHueYq4nuR6dcuKnAAp6SicssUXiFfZGCW/POgSQEN1q5tvZn0jXjnxFG2fqYNoZsGxcRciKAisuF9hS03Aq57DcwwTCjbt/BVXLFNuz9Yolr+ROkoukvZDhEbnADqZFzXyCWOMv8WcihpH/7h9YCMHVVsDnidMeJZwSMGjY139l+C1SAXEtzRH5LWuUKE0QGgQnpjjKngVnp6FeMwL3+ex10bTt2lJ1cZrZO7Hj5DzPn//JprOPFFUd3Y//3AGRcm49\\"></Mermaid>"`;

exports[`mermaid > Should render \`\`\`state 1`] = `"<Mermaid id=\\"mermaid-0\\" code=\\"eJwrLkksSXXJTEwvSszVLTPiUlCI1opV0NW1U3BMLsksS+UCioDVQPkK1UABhCK/0lyf/ORs/7Q0sDCCiyKbp2Cl4FoG5QUUpRYXp6agqM9DMwyXel1dFNudEwuKka1H4qPKQxwA4yKbiKQEzUScWtAcEZxclJ+Tg+wMFBF0NRCnIASQTUZRhmE2To21XAApZ4wz\\"></Mermaid>"`;
